File: daily/Binary Search/1300-Sum-of-mutated-array-closest-to-target.py
from typing import List


class Solution:
    def findBestValue(self, arr: List[int], target: int) -> int:
        max_value = max(arr)
        low, high = 1, max_value

        while low < high:
            mid = (low + high) // 2
            if self._mutated_sum(arr, mid) <= target:
                low = mid + 1
            else:
                high = mid

        sum_low = self._mutated_sum(arr, low)
        sum_prev = self._mutated_sum(arr, low - 1)
        if abs(sum_low - target) < abs(sum_prev - target):
            return low
        return low - 1

    def _mutated_sum(self, arr: List[int], value: int) -> int:
        total = 0
        for num in arr:
            total += value if num > value else num
        return total

File: daily/Binary Search/1300-Sum-of-mutated-array-closest-to-target.go
package daily

func findBestValue(arr []int, target int) int {
	maxValue := 0
	for _, val := range arr {
		if val > maxValue {
			maxValue = val
		}
	}
	low, high := 1, maxValue
	for low < high {
		mid := (low + high) / 2
		if sumAtMid(arr, mid) <= target {
			low = mid + 1
		} else {
			high = mid
		}
	}

	s1 := sumAtMid(arr, low)
	s2 := sumAtMid(arr, low-1)
	if abs(s1-target) < abs(s2-target) {
		return low
	}
	return low - 1
}

func sumAtMid(arr []int, value int) int {
	sum := 0
	for _, x := range arr {
		if x > value {
			sum += value
		} else {
			sum += x
		}
	}
	return sum
}
